Add nginx, apache and caddy webserver log parsers

JSCU-CNI opened the pull request:
This PR adds support for nginx, apache and caddy webserver log parsing. This PR also moves the existing IIS parser to the `plugins/apps/webservers` namespace.

Schamper commented:
I've not looked at this PR extensively yet, but it looks like most comments regarding timestamps and reading compressed files from #131 are also applicable to this PR. Could you address those first?

JSCU-CNI commented:
Implemented comments from #131. Please note that https://github.com/fox-it/dissect.target/pull/137 interferes with this PR since we changed the location of `iis` to `plugins/apps/webservers/iis.py`.

Schamper commented:
Thanks, yeah we'll need to make sure we merge that in the correct order. Also reminder for ourselves that this requires a change in [acquire](https://github.com/fox-it/acquire/blob/e8f05f3c91c107269b61f4cb517a3bafd4634d4d/acquire/acquire.py#L22) as well.

Schamper commented on dissect/target/plugins/apps/webservers/apache.py:
```diff
@@ -0,0 +1,126 @@
+LOG_FORMATS = {
+    "vhost_combined": re.compile(rf"(?P<server_name>.*?):(?P<port>.*) {COMMON_REGEX} {REFERER_USER_AGENT_REGEX}"),
+    "combined": re.compile(rf"{COMMON_REGEX} {REFERER_USER_AGENT_REGEX}"),
+    "common": re.compile(rf"{COMMON_REGEX}"),
```
```suggestion
    "common": re.compile(COMMON_REGEX),
```

Schamper commented on dissect/target/plugins/apps/webservers/webservers.py:
```diff
@@ -0,0 +1,56 @@
+        ("net.ipaddress", "ipaddr"),
+        ("uri", "file_path"),
+        ("wstring", "url"),
+        ("varint", "statuscode"),
+        ("varint", "bytessent"),
```
```suggestion
        ("varint", "status_code"),
        ("varint", "bytes_sent"),
```
?

Schamper commented on dissect/target/plugins/apps/webservers/webservers.py:
```diff
@@ -0,0 +1,56 @@
+        ("datetime", "ts"),
+        ("string", "remote_user"),
+        ("net.ipaddress", "ipaddr"),
+        ("uri", "file_path"),
```
```suggestion
        ("path", "source"),
```
And move to the end of the descriptor.

Schamper commented on tests/test_plugins_apps_webservers_apache.py:
```diff
@@ -0,0 +1,109 @@
+from ._utils import absolute_path
+
+
+class TestHelpers:
```
Can you change these to just be regular test functions.

Schamper commented on tests/test_plugins_apps_webservers_apache.py:
```diff
@@ -0,0 +1,109 @@
+        assert infer_log_format(log_common) == "combined"
+
+
+class TestApache:
```
Can you change these to just be regular test functions.

Schamper commented on tests/test_plugins_apps_webservers_nginx.py:
```diff
@@ -0,0 +1,64 @@
+from ._utils import absolute_path
+
+
+class TestNginx:
```
Can you change these to just be regular test functions.

Schamper commented on dissect/target/plugins/apps/webservers/webservers.py:
```diff
@@ -0,0 +1,56 @@
+                self.target.log.exception(f"Failed to execute tool plugin: {p._name}{f}")
+
+    @export(record=WebserverRecord)
+    def history(self):
```
Would `access` not be more accurate?

JSCU-CNI replied:
I agree, `logs` or `access` would be better. Which do you prefer? I think `iis` uses `logs`.

Schamper replied:
I think `access` would be best for these currently. `logs` could be added later to be a sort of shorthand that calls the other log plugins (e.g. error).

JSCU-CNI replied:
Added in [3106dde](https://github.com/fox-it/dissect.target/pull/127/commits/3106dde91dc2b556bb867dde4cee456e9fa374c3)

Schamper commented on dissect/target/plugins/apps/webservers/nginx.py:
```diff
@@ -0,0 +1,81 @@
+
+        logs_dir = self.target.fs.path("/var/log/nginx")
+        if logs_dir.exists():
+            return [self.target.fs.path(p) for p in logs_dir.glob("access.log*")]
```
Would you not rather combine this with possible output of parsing the config file?

Schamper commented on dissect/target/plugins/apps/webservers/nginx.py:
```diff
@@ -0,0 +1,81 @@
+    def history(self):
+        for path in self.get_log_paths():
+            log_lines = [line.strip() for line in open_decompress(path, "rt")]
+            yield from self.parse_nginx_logs(log_lines, path)
```
Same thing here, I think you're better off removing a couple of methods and moving them into this method.

Schamper commented on dissect/target/plugins/apps/webservers/caddy.py:
```diff
@@ -0,0 +1,84 @@
+    @plugin.export(record=WebserverRecord)
+    def history(self):
+        log_path = self.get_log_path()
+        yield from self.parse_caddy_logs(log_path)
```
Same thing here, I think you're better off removing a couple of methods and moving them into this method.

Schamper commented on dissect/target/plugins/apps/webservers/caddy.py:
```diff
@@ -0,0 +1,84 @@
+        log_file = "/var/log/caddy_access.log"
+        if log_file.exists():
+            self.target.log.debug(f"Log file found at {self.LOG_FILE_PATH}")
+            return log_file
```
Would you not rather combine this with possible output of parsing the config file?

Schamper commented on dissect/target/plugins/apps/webservers/apache.py:
```diff
@@ -0,0 +1,111 @@
+                    )
+                    continue
+
+                match = match.groupdict()
+                yield WebserverRecord(
+                    ts=parse_datetime(match.get("ts"), tzinfo),
```
```suggestion
                match = match.groupdict()
                yield WebserverRecord(
                    ts=datetime.strptime(match.get("ts"), "%d/%b/%Y:%H:%M:%S %z").replace(tzinfo=tzinfo),
```

Schamper commented on dissect/target/plugins/apps/webservers/apache.py:
```diff
@@ -0,0 +1,111 @@
+    return None
+
+
+def parse_datetime(ts: str, tz: ZoneInfo) -> datetime:
+    return datetime.strptime(ts, "%d/%b/%Y:%H:%M:%S %z").replace(tzinfo=tz)
+
+
```
```suggestion
```

Schamper commented on dissect/target/plugins/apps/webservers/apache.py:
```diff
@@ -0,0 +1,111 @@
+
+    def __init__(self, target):
+        super().__init__(target)
+        self.LOGS_DIR_PATH = self.get_log_paths()
```
```suggestion
        self.log_paths = self.get_log_paths()
```

Schamper commented on dissect/target/plugins/apps/webservers/apache.py:
```diff
@@ -0,0 +1,111 @@
+        return list(self.target.fs.path(default_logs_dir).glob("access.log*"))
+
+    def check_compatible(self):
+        return len(self.LOGS_DIR_PATH) > 0
```
```suggestion
        return len(self.log_paths) > 0
```

Schamper commented on dissect/target/plugins/apps/webservers/apache.py:
```diff
@@ -0,0 +1,111 @@
+    def logs(self):
+        tzinfo = self.target.datetime.tzinfo
+
+        for path in self.LOGS_DIR_PATH:
```
```suggestion
        for path in self.log_paths:
```

Schamper commented on dissect/target/plugins/apps/webservers/apache.py:
```diff
@@ -0,0 +1,111 @@
+        return len(self.LOGS_DIR_PATH) > 0
+
+    @plugin.export(record=WebserverRecord)
+    def logs(self):
```
```suggestion
    def access(self) -> Iterator[WebserverRecord]:
```

Schamper commented on dissect/target/plugins/apps/webservers/caddy.py:
```diff
@@ -0,0 +1,77 @@
+
+    def __init__(self, target):
+        super().__init__(target)
+        self.LOG_FILE_PATH = self.get_log_paths()
```
```suggestion
        self.log_paths = self.get_log_paths()
```

Schamper commented on dissect/target/plugins/apps/webservers/caddy.py:
```diff
@@ -0,0 +1,77 @@
+        return self.target.fs.path(self.LOG_FILE_PATH).exists()
+
+    @plugin.export(record=WebserverRecord)
+    def logs(self):
```
```suggestion
    def access(self) -> Iterator[WebserverRecord]:
```

Schamper commented on dissect/target/plugins/apps/webservers/caddy.py:
```diff
@@ -0,0 +1,77 @@
+
+        return log_paths
+
+    def check_compatible(self):
```
```suggestion
    def check_compatible(self) -> bool:
```

Schamper commented on dissect/target/plugins/apps/webservers/caddy.py:
```diff
@@ -0,0 +1,77 @@
+        self.LOG_FILE_PATH = self.get_log_paths()
+
+    @plugin.internal
+    def get_log_paths(self):
```
```suggestion
    def get_log_paths(self) -> list[Path]:
```

Schamper commented on dissect/target/plugins/apps/webservers/caddy.py:
```diff
@@ -0,0 +1,77 @@
+
+    __namespace__ = "caddy"
+
+    def __init__(self, target):
```
```suggestion
    def __init__(self, target: Target):
```

Schamper commented on dissect/target/plugins/apps/webservers/caddy.py:
```diff
@@ -0,0 +1,77 @@
+)
+
+
+def parse_datetime(date_str: str, tz: ZoneInfo):
+    # Example: 10/Apr/2020:14:10:12 +0000
+    return datetime.strptime(f"{date_str}", "%d/%b/%Y:%H:%M:%S %z").replace(tzinfo=tz)
+
+
```
```suggestion
```

Schamper commented on dissect/target/plugins/apps/webservers/caddy.py:
```diff
@@ -0,0 +1,77 @@
+            for line in lines:
+                line = line.strip()
+                if "log  " in line:
+                    log_paths.append(self.target.fs.path(line.split(" ")[1]).parent)
```
Maybe check if it exists?

Schamper commented on dissect/target/plugins/apps/webservers/caddy.py:
```diff
@@ -0,0 +1,77 @@
+        """
+        tzinfo = self.target.datetime.tzinfo
+
+        for path in self.LOGS_DIR_PATH:
```
```suggestion
        for path in self.log_paths:
```

Schamper commented on dissect/target/plugins/apps/webservers/caddy.py:
```diff
@@ -0,0 +1,77 @@
+        tzinfo = self.target.datetime.tzinfo
+
+        for path in self.LOGS_DIR_PATH:
+            log_lines = [line.strip() for line in open_decompress(path, "rt")]
+
+            for line in log_lines:
```
```suggestion
            for line in open_decompress(path, "rt"):
                line = line.strip()
                if not line:
                    continue
```

Schamper commented on dissect/target/plugins/apps/webservers/caddy.py:
```diff
@@ -0,0 +1,77 @@
+
+                match = match.groupdict()
+                yield WebserverRecord(
+                    ts=parse_datetime(match.get("datetime"), tzinfo),
```
```suggestion
                    ts=datetime.strptime(match.get("datetime"), "%d/%b/%Y:%H:%M:%S %z").replace(tzinfo=tzinfo),
```

Schamper commented on dissect/target/plugins/apps/webservers/caddy.py:
```diff
@@ -0,0 +1,77 @@
+                    url=match.get("url"),
+                    status_code=match.get("status_code"),
+                    bytes_sent=match.get("bytes_sent"),
+                    source=path.resolve().__str__(),
```
```suggestion
                    source=path.resolve(),
```

Schamper commented on dissect/target/plugins/apps/webservers/apache.py:
```diff
@@ -0,0 +1,111 @@
+                    bytes_sent=match.get("bytes_sent"),
+                    referer=match.get("referer"),
+                    useragent=match.get("useragent"),
+                    source=path.resolve().__str__(),
```
```suggestion
                    source=path.resolve(),
```

Schamper commented on dissect/target/plugins/apps/webservers/iis.py:
```diff
@@ -269,6 +267,10 @@ def logs(self) -> Generator[TargetRecordDescriptor, None, None]:
             yield from parse_func(log_file)
 
+    @plugin.export(record=BasicRecordDescriptor)
+    def history(self) -> Generator[TargetRecordDescriptor, None, None]:
```
```suggestion
    def access(self) -> Generator[TargetRecordDescriptor, None, None]:
```

Schamper commented on dissect/target/plugins/apps/webservers/nginx.py:
```diff
@@ -0,0 +1,75 @@
+    def __init__(self, target):
+        super().__init__(target)
+        self.LOGS_DIR_PATH = self.get_log_paths()
+        self.target_timezone = target.datetime.tzinfo
```
```suggestion
```

Schamper commented on dissect/target/plugins/apps/webservers/nginx.py:
```diff
@@ -0,0 +1,75 @@
+
+    def __init__(self, target):
+        super().__init__(target)
+        self.LOGS_DIR_PATH = self.get_log_paths()
```
```suggestion
        self.log_paths = self.get_log_paths()
```

Schamper commented on dissect/target/plugins/apps/webservers/nginx.py:
```diff
@@ -0,0 +1,75 @@
+class NginxPlugin(plugin.Plugin):
+    __namespace__ = "nginx"
+
+    def __init__(self, target):
```
```suggestion
    def __init__(self, target: Target):
```

Schamper commented on dissect/target/plugins/apps/webservers/apache.py:
```diff
@@ -0,0 +1,111 @@
+        default_logs_dir = "/var/log/apache2"
+        return list(self.target.fs.path(default_logs_dir).glob("access.log*"))
+
+    def check_compatible(self):
```
```suggestion
    def check_compatible(self) -> bool:
```

Schamper commented on dissect/target/plugins/apps/webservers/apache.py:
```diff
@@ -0,0 +1,111 @@
+        self.LOGS_DIR_PATH = self.get_log_paths()
+
+    @plugin.internal
+    def get_log_paths(self) -> []:
```
```suggestion
    def get_log_paths(self) -> list[Path]:
```

Schamper commented on dissect/target/plugins/apps/webservers/apache.py:
```diff
@@ -0,0 +1,111 @@
+
+    __namespace__ = "apache"
+
+    def __init__(self, target):
```
```suggestion
    def __init__(self, target: Target):
```

Schamper commented on dissect/target/plugins/apps/webservers/nginx.py:
```diff
@@ -0,0 +1,75 @@
+)
+
+
+def parse_datetime(date_str: str, tz: ZoneInfo):
+    # Example: 10/Apr/2020:14:10:12 +0000
+    return datetime.strptime(f"{date_str}", "%d/%b/%Y:%H:%M:%S %z").replace(tzinfo=tz)
+
+
```
```suggestion
```

Schamper commented on dissect/target/plugins/apps/webservers/nginx.py:
```diff
@@ -0,0 +1,75 @@
+        self.target_timezone = target.datetime.tzinfo
+
+    @plugin.internal
+    def get_log_paths(self):
```
```suggestion
    def get_log_paths(self) -> list[Path]:
```

Schamper commented on dissect/target/plugins/apps/webservers/nginx.py:
```diff
@@ -0,0 +1,75 @@
+            for line in lines:
+                line = line.strip()
+                if "access_log" in line:
+                    log_paths.append(self.target.fs.path(line.split(" ")[1]).parent)
```
Maybe check if it exists?

Schamper commented on dissect/target/plugins/apps/webservers/nginx.py:
```diff
@@ -0,0 +1,75 @@
+
+        # Check for custom paths in nginx install config
+        if (config_file := self.target.fs.path("/etc/nginx/nginx.conf")).exists():
+            lines = config_file.open("rt").readlines()
+            for line in lines:
```
```suggestion
            for line in config_file.open("rt"):
```

Schamper commented on dissect/target/plugins/apps/webservers/caddy.py:
```diff
@@ -0,0 +1,77 @@
+
+        # Check for custom paths in Caddy config
+        if (config_file := self.target.fs.path("/etc/caddy/Caddyfile")).exists():
+            lines = config_file.open("rt").readlines()
+            for line in lines:
```
```suggestion
            for line in config_file.open("rt"):
```

Schamper commented on dissect/target/plugins/apps/webservers/nginx.py:
```diff
@@ -0,0 +1,75 @@
+
+        return log_paths
+
+    def check_compatible(self):
```
```suggestion
    def check_compatible(self) -> bool:
```

Schamper commented on dissect/target/plugins/apps/webservers/nginx.py:
```diff
@@ -0,0 +1,75 @@
+        return log_paths
+
+    def check_compatible(self):
+        return len(self.LOGS_DIR_PATH) > 0
```
```suggestion
        return len(self.log_paths) > 0
```

Schamper commented on dissect/target/plugins/apps/webservers/nginx.py:
```diff
@@ -0,0 +1,75 @@
+        return len(self.LOGS_DIR_PATH) > 0
+
+    @plugin.export(record=WebserverRecord)
+    def logs(self):
```
```suggestion
    def access(self) -> Iterator[WebserverRecord]:
```

Schamper commented on dissect/target/plugins/apps/webservers/nginx.py:
```diff
@@ -0,0 +1,75 @@
+    def logs(self):
+        tzinfo = self.target.datetime.tzinfo
+
+        for path in self.LOGS_DIR_PATH:
```
```suggestion
        for path in self.log_paths:
```

Schamper commented on dissect/target/plugins/apps/webservers/nginx.py:
```diff
@@ -0,0 +1,75 @@
+        tzinfo = self.target.datetime.tzinfo
+
+        for path in self.LOGS_DIR_PATH:
+            log_lines = [line.strip() for line in open_decompress(path, "rt")]
+
+            for line in log_lines:
```
```suggestion
            for line in open_decompress(path, "rt"):
                line = line.strip()
                if not line:
                    continue
```

Schamper commented on dissect/target/plugins/apps/webservers/nginx.py:
```diff
@@ -0,0 +1,75 @@
+
+                match = match.groupdict()
+                yield WebserverRecord(
+                    ts=parse_datetime(match.get("datetime"), tzinfo),
```
```suggestion
                    ts=datetime.strptime(match.get("datetime"), "%d/%b/%Y:%H:%M:%S %z").replace(tzinfo=tzinfo),
```

Schamper commented on dissect/target/plugins/apps/webservers/nginx.py:
```diff
@@ -0,0 +1,75 @@
+                    bytes_sent=match.get("bytes_sent"),
+                    referer=match.get("referer"),
+                    useragent=match.get("useragent"),
+                    source=path.resolve().__str__(),
```
```suggestion
                    source=path.resolve(),
```

Schamper commented on dissect/target/plugins/apps/webservers/webservers.py:
```diff
@@ -0,0 +1,62 @@
+        ("datetime", "ts"),
+        ("string", "remote_user"),
+        ("net.ipaddress", "remote_ip"),
+        ("wstring", "url"),
```
```suggestion
        ("uri", "url"),
```

Schamper commented on dissect/target/plugins/apps/webservers/webservers.py:
```diff
@@ -0,0 +1,62 @@
+        ("wstring", "url"),
+        ("varint", "status_code"),
+        ("varint", "bytes_sent"),
+        ("wstring", "referer"),
```
```suggestion
        ("uri", "referer"),
```

Schamper commented on dissect/target/plugins/apps/webservers/webservers.py:
```diff
@@ -0,0 +1,62 @@
+        "caddy",
+    ]
+
+    def __init__(self, target):
```
```suggestion
    def __init__(self, target: Target):
```

JSCU-CNI commented:
Implemented all suggestions in [304280b](https://github.com/fox-it/dissect.target/pull/127/commits/304280be38ce03df7bbda25d3f780cd48a43bdfc) and [eed2a5b](https://github.com/fox-it/dissect.target/pull/127/commits/eed2a5b778b41bf8c1d7933dba81df3388eac363).

Schamper commented on dissect/target/plugins/apps/webservers/caddy.py:
```diff
@@ -0,0 +1,74 @@
+            log_paths.append(default_log_file)
+
+        # Check for custom paths in Caddy config
+        if (config_file := self.target.fs.path("/etc/caddy/Caddyfile")).exists():
```
Can you add a unit test for parsing the log file?

JSCU-CNI replied:
added in [760128a](https://github.com/fox-it/dissect.target/pull/127/commits/760128a6843d771e324b285adcfd3cc4368ea635)

Schamper commented on dissect/target/plugins/apps/webservers/caddy.py:
```diff
@@ -0,0 +1,74 @@
+        return self.target.fs.path(self.LOG_FILE_PATH).exists()
+
+    @plugin.export(record=WebserverRecord)
+    def access(self) -> Iterator[WebserverRecord]:
```
Can you add a unit test for caddy log files?

JSCU-CNI replied:
added in [760128a](https://github.com/fox-it/dissect.target/pull/127/commits/760128a6843d771e324b285adcfd3cc4368ea635)

Schamper commented on dissect/target/plugins/apps/webservers/nginx.py:
```diff
@@ -0,0 +1,72 @@
+            log_paths = [self.target.fs.path(p) for p in default_logs_dir.glob("access.log*")]
+
+        # Check for custom paths in nginx install config
+        if (config_file := self.target.fs.path("/etc/nginx/nginx.conf")).exists():
```
Can you add a unit test for parsing nginx log files?

JSCU-CNI replied:
Added in [58c49a2](https://github.com/fox-it/dissect.target/pull/127/commits/58c49a2a053d4fe28b4d10ed4bd3b449c0e98219).

Schamper commented on dissect/target/plugins/apps/webservers/caddy.py:
```diff
@@ -0,0 +1,128 @@
+                    p = match["log_file"].replace(" {", "").split(" ")[-1]
+
+                    if not p.startswith("/"):
+                        print(f"{p} is a relative path")
```
```suggestion
```

Schamper commented on dissect/target/plugins/apps/webservers/caddy.py:
```diff
@@ -0,0 +1,128 @@
+                        for root in found_roots:
+                            root_parent = self.target.fs.path(f"{root}").parent
+                            abs_log_path = self.target.fs.path(f"{str(root_parent)}/{p}").parent
+                            print(abs_log_path)
```
```suggestion
```

Schamper commented on dissect/target/plugins/apps/webservers/caddy.py:
```diff
@@ -0,0 +1,128 @@
+                            if abs_log_path.exists() and abs_log_path not in log_paths:
+                                log_paths.append(abs_log_path)
+                    else:
+                        print(f"{p} is an absolute path")
```
```suggestion
```

Schamper commented on dissect/target/plugins/apps/webservers/caddy.py:
```diff
@@ -0,0 +1,128 @@
+                    if not p.startswith("/"):
+                        print(f"{p} is a relative path")
+                        for root in found_roots:
+                            root_parent = self.target.fs.path(f"{root}").parent
```
```suggestion
                            root_parent = self.target.fs.path(root).parent
```

Schamper commented on dissect/target/plugins/apps/webservers/caddy.py:
```diff
@@ -0,0 +1,128 @@
+                        print(f"{p} is a relative path")
+                        for root in found_roots:
+                            root_parent = self.target.fs.path(f"{root}").parent
+                            abs_log_path = self.target.fs.path(f"{str(root_parent)}/{p}").parent
```
```suggestion
                            abs_log_path = self.target.fs.path(root_parent).joinpath(p).parent
```

Schamper commented on dissect/target/plugins/apps/webservers/caddy.py:
```diff
@@ -0,0 +1,128 @@
+                    yield WebserverRecord(
+                        ts=from_unix(log["ts"]).replace(tzinfo=tzinfo),
+                        remote_ip=log["request"]["remote_ip"],
+                        url=f"{log['request']['method']} {log['request']['uri']} {log['request']['proto']}",
```
I assume `log['request']['method']` is strings like `GET`, `POST` etc, and `log['request']['proto']` is `HTTP/2`?

I'd suggest putting just the `log['request']['uri']` in the `url` field of the record, and either create new fields for the `proto` and `method` if you want to include them.

JSCU-CNI replied:
This is to ensure uniformity across all other webserver log parsers. The format for nginx, apache and caddy CLF is always `method uri proto`.

Schamper replied:
I changed it to separate `method`, `uri` and `protocol` fields.

Schamper commented on dissect/target/plugins/apps/webservers/caddy.py:
```diff
@@ -0,0 +1,128 @@
+                    log = match.groupdict()
+
+                    yield WebserverRecord(
+                        ts=datetime.strptime(log["ts"], "%d/%b/%Y:%H:%M:%S %z").replace(tzinfo=tzinfo),
```
```suggestion
                        ts=datetime.strptime(log["ts"], "%d/%b/%Y:%H:%M:%S %z"),
```
I notice a `%z` in the `strptime`, so that should already parse this log line to the correct timezone.

Schamper commented on dissect/target/plugins/apps/webservers/caddy.py:
```diff
@@ -0,0 +1,128 @@
+                        continue
+
+                    yield WebserverRecord(
+                        ts=from_unix(log["ts"]).replace(tzinfo=tzinfo),
```
```suggestion
                        ts=from_unix(log["ts"]),
```
I assume `Caddy` logs true Unix (i.e. UTC) timestamps in this field? So no need to transform to another timezone.

JSCU-CNI replied:
Unfortunately this does not seem to be the case. The epoch timestamp seems to be relative to the system's time.
```
FAILED tests/test_plugins_apps_webservers_caddy.py::test_plugins_apps_webservers_caddy_json - AssertionError: assert 2023-02-06 15:05:49.064393+00:00 == 2023-02-06 15:05:49.064393+01:00
```

JSCU-CNI replied:
Re-added in [4e5da15](https://github.com/fox-it/dissect.target/pull/127/commits/4e5da1558e3db1b9c9c76ac1188422b3442c4c71).

Schamper commented on dissect/target/plugins/apps/webservers/nginx.py:
```diff
@@ -0,0 +1,71 @@
+            for line in config_file.open("rt"):
+                line = line.strip()
+                if "access_log " in line:
+                    p = list(filter(None, line.strip().split(" ")))[1]
```
```suggestion
                    p = line.split()[1]
```
I'm not sure if this was protective coding against something specific, but I believe this should achieve the same thing.

Also probably good to be a bit more defensive here, check if the result of `line.split()` is at least 2.

JSCU-CNI replied:
This is to ensure we correctly parse all possible `access_log` definitions (see tests/data/webservers/nginx/nginx.conf).

Schamper replied:
I changed it to `line.split()` and the unit tests still passed.

Schamper commented on dissect/target/plugins/apps/webservers/nginx.py:
```diff
@@ -0,0 +1,71 @@
+                line = line.strip()
+                if "access_log " in line:
+                    p = list(filter(None, line.strip().split(" ")))[1]
+                    if (path := self.target.fs.path(p).parent).exists():
```
Won't accessing the `.parent` here give you the directory where the log file is located? There's no iterating over any files happening after this.

Maybe a good idea to change `get_log_paths` to instead be `get_log_dirs` and do the glob for `access.log*` in the parsing function below. Otherwise you're doing the same glob in multiple places.

Same comment for the Caddy plugin by the way.

JSCU-CNI replied:
Good catch. Improved this behaviour in [01412a9](https://github.com/fox-it/dissect.target/pull/127/commits/01412a927a93b4751bd6abaae3cbba0bcc66802c).

Schamper commented on dissect/target/plugins/apps/webservers/nginx.py:
```diff
@@ -0,0 +1,71 @@
+
+                match = match.groupdict()
+                yield WebserverRecord(
+                    ts=datetime.strptime(match["datetime"], "%d/%b/%Y:%H:%M:%S %z").replace(tzinfo=tzinfo),
```
```suggestion
                    ts=datetime.strptime(match["datetime"], "%d/%b/%Y:%H:%M:%S %z"),
```
Same comment here, `%z` in the `strptime` should already parse with the correct timestamp information.

JSCU-CNI commented:
Some things were added on top of the implemented review feedback:
- Modify `iis.access()` to behave like the other webserver plugins. [b6a38bf](https://github.com/fox-it/dissect.target/pull/127/commits/b6a38bfabb31c166df11e478b41260c323eb7440)
- Add more default log paths to the apache plugin. [01412a9](https://github.com/fox-it/dissect.target/pull/127/commits/01412a927a93b4751bd6abaae3cbba0bcc66802c)
- All webserver log parsers introduced in this PR (nginx, apache, caddy) should now always detect logrotated files as well (eg. `access.log`, `access.log.1`, etc.)

Schamper commented on dissect/target/plugins/apps/webservers/apache.py:
```diff
@@ -0,0 +1,154 @@
+
+                    log = match.groupdict()
+                    yield WebserverAccessLogRecord(
+                        ts=datetime.strptime(log["ts"], "%d/%b/%Y:%H:%M:%S %z").replace(tzinfo=tzinfo),
```
Yes, my bad, I forgot that one. Also removed some other leftover code.

JSCU-CNI commented:
Thanks for reviewing and the suggestions. Looks good to me :)

Schamper commented on dissect/target/plugins/apps/webservers/iis.py:
```diff
@@ -267,6 +279,29 @@ def logs(self) -> Generator[TargetRecordDescriptor, None, None]:
+                remote_ip=iis_record.client_ip,
+                method=iis_record.request_method,
+                uri=iis_record.request_path,
+                protocol=getattr(iis_record, "cs_version", None),
```
`cs_version` is not guaranteed to be available on the records generated by the IIS plugin.

Schamper replied:
Oversight, I didn't notice the other ones were using `_asdict`.

Schamper commented on dissect/target/plugins/apps/webservers/nginx.py:
```diff
@@ -0,0 +1,82 @@
+        return len(self.log_paths) > 0
+
+    @plugin.export(record=WebserverAccessLogRecord)
+    def access(self) -> Iterator[WebserverAccessLogRecord]:
```
I thought about this but personally found it a bit overkill, especially since there are slight differences that would quickly make this quite complex. The shared logic is small enough that I felt it was okay for now.

JSCU-CNI commented:
Added a last-minute improvement in [cee0024](https://github.com/fox-it/dissect.target/pull/127/commits/cee00248c0561b30c53d1a81af3a798c1b3048c6) based on @Schamper's comment in https://github.com/fox-it/dissect.target/pull/173#discussion_r1106789286.

Schamper commented on dissect/target/plugins/apps/webservers/caddy.py:
```diff
@@ -0,0 +1,138 @@
+            found_roots = []
+            for line in config_file.open("rt"):
+                line = line.strip()
+                if not line or line.startswith("#"):
```
What about Caddy, can you also fix it here? Seems a bit more complicated unfortunately.

JSCU-CNI replied:
Added in [bd39b7a](https://github.com/fox-it/dissect.target/pull/127/commits/bd39b7a4ec215638055a802d5044797d027c3bd3).

Schamper replied:
Slightly changed it :)
